src/Commands.hs

```diff
@@ -342,7 +342,11 @@ commandBuild ctx [XObj (Bol shutUp) _ _] = do
         if generateOnly
           then pure (ctx, dynamicNil)
           else case Map.lookup "main" (envBindings env) of
-            Just _ -> compile True
+            Just (Binder _ (XObj _ _ (Just (FuncTy _ IntTy _)))) -> compile True
+            Just (Binder _ (XObj _ _ (Just (FuncTy _ UnitTy _)))) -> compile True
+            Just (Binder _ (XObj _ _ (Just (FuncTy _ _ _)))) -> pure (evalError ctx "`main` was generic at build time, it must return either `Int` or `()`." Nothing)
```

Comment:
nit: the return type here could be non-generic, e.g. `String`. Maybe change the language to:

``main` did not return an Int or () at build time. It must return either `Int` or `()` but returns: show retT`

Reply:
The non-generic case is actually already handled, as detailed in #1246, by the concretization error.

Reply:
Ah, right—that makes sense. I worry that it's easy to forget that prior checks are performed  before we reach this code, which might lead someone to duplicate the code by adding an additional check (at least that's what I would have done if you didn't point this out!). Maybe we can add a one-line comment explaining this for now? Likely this indicates we should eventually move the check into concretize...but that's not great either imo, since I don't think concretize should really care that a function happens to be called `main`—that feels like we're sliding into ambiguous separation of concerns. Maybe all the checks that are in concretize should be moved here.

Reply:
That’s a good point—I actually just checked and saw that concretization had a check for generic types in `main`. Removing that check sadly breaks the compiler, since it enforces that concretization happens in a single pass, essentially. So I think the problem here is that concretization can‘t enforce specialization, which is why we need the special cases in the build commands. Catching what we can as early as possible (and concretization happens before building) makes sense to me, though, especially in the REPL. What do you think?

Thanks for the input!

Reply:
I did add a comment as you suggested, btw!

Reply:
Ah, gotcha—catching early makes sense too. Comment looks great, thank you!